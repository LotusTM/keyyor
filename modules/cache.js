const LRUCache = require('lru-cache')
const hash = require('hash-sum')

const cache = new LRUCache()

/**
 * Cache function into specified cache key or based on hash of its arguments
 * @param  {function} func       Function to be cached
 * @param  {string}   [cacheKey] Optional cache store key to store result.
 *                               If store already has `cacheKey` value, it will be
 *                               immediately returned
 * @return {function} Cached function to be called with desired arguments
 * @example
 *  cacheFunc(renderFunc)('{{ value }}', { value: 'test' })
 *  cacheFunc(renderFunc, 'matter')('{{ value }}', { value: 'test2' })
 */
const cacheFunc = (func, cacheKey) => (...args) => {
  const key = cacheKey || hash(args)
  const cached = cache.get(key)

  if (cached) return cached

  const result = func(...args)

  cache.set(key, result)

  return result
}

module.exports = {
  cache,
  cacheFunc
}
