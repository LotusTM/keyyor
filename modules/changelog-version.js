const { exec } = require('child_process')
const { readFile, writeFileSync } = require('fs')
const { join } = require('path')
const moment = require('moment')
const pkg = require(join(process.cwd(), 'package.json'))
const { green, red } = require('chalk')

const isFromCLI = require.main === module
const logError = (message) => console.error(red(message))

const VERSION = pkg.version
const CHANGELOG_FILENAME = 'CHANGELOG.md'
const REPOSITORY_URL = pkg.repository.url
const DATE = moment().format('YYYY-MM-DD')
const HEAD_PATTERN = /^## \[?(HEAD|Unreleased)\]?.+$/m
const VERSION_PATTERN = /^## \[?(\d+\.\d+\.\d+)/m

/**
 * Update inputted changelog version
 *
 * @param  {object} [options]                   Options
 * @param  {string} [options.input]             Input to be changed
 * @param  {string} [options.version]           Version to be used
 * @param  {string} [options.repositoryURL]     URL to the project repository
 * @param  {string} [options.date]              Version date
 * @param  {RegExp} [options.headPattern]       Pattern to match HEAD heading to update
 * @param  {RegExp} [options.versionPattern]    Pattern to match versions in changelog
 * @return {string} Updated input
 */
const updateVersion = ({
  input,
  version = VERSION,
  repositoryURL = REPOSITORY_URL,
  date = DATE,
  headPattern = HEAD_PATTERN,
  versionPattern = VERSION_PATTERN
} = {}) => {
  if (!version) {
    logError('[changelog-version] please, specify `package.json` `version` property')
    process.exit(1)
  }

  if (!repositoryURL || !repositoryURL.includes('github.com')) {
    logError('[changelog-version] please, specify `package.json` `repository.url` property with a valid Github URL')
    process.exit(1)
  }

  if (repositoryURL.endsWith('.git')) {
    repositoryURL = repositoryURL.slice(0, -4)
  }

  const hasHead = headPattern.test(input)

  if (!hasHead) {
    logError('[changelog-version] changelog does not have HEAD version, can not update it with latest version')
    process.exit(1)
  }

  const previousVersion = input.match(versionPattern)[1]
  const releaseHeader = `## [${version}](${repositoryURL}/compare/v${previousVersion}...v${version}) - ${date}`
  const newHEADHeader = `## [HEAD](${repositoryURL}/compare/v${version}...HEAD)`

  return input.replace(headPattern, `${newHEADHeader}\n\n${releaseHeader}`)
}

/**
 * Update `CHANGELOG.md` HEAD to the latest version,
 * with proper compare URL and date, as well as
 * add new `HEAD` header with proper URL.
 *
 * CHANGELOG should follow [keepchangelog.com specs](https://keepachangelog.com).
 *
 * Relies on `package.json` `version` and `repository.url` values.
 *
 * Sould be run in `version` NPM script.
 *
 * @param  {object} [options]                   Options
 * @param  {string} [options.version]           Version to be used
 * @param  {string} [options.changelogFilename] Changelog filename
 * @param  {string} [options.repositoryURL]     URL to the project repository
 * @param  {string} [options.date]              Version date
 * @param  {RegExp} [options.headPattern]       Pattern to match HEAD heading to update
 * @param  {RegExp} [options.versionPattern]    Pattern to match versions in changelog
 * @return {void} It just updates `CHANGELOG.md`
 */
const updateChangelogVersion = ({
  version = VERSION,
  changelogFilename = CHANGELOG_FILENAME,
  repositoryURL = REPOSITORY_URL,
  date = DATE,
  headPattern = HEAD_PATTERN,
  versionPattern = VERSION_PATTERN
} = {}) => readFile(changelogFilename, { encoding: 'utf-8' }, (error, changelog) => {
  if (error) return console.log(`[changelog-version] seems to be no ${changelogFilename}, skipping changelog version update`)

  writeFileSync(changelogFilename, updateVersion({
    input: changelog,
    version,
    changelogFilename,
    repositoryURL,
    date,
    headPattern,
    versionPattern
  }))

  exec(`git add ${changelogFilename}`, (error, stdout) => {
    if (error) {
      logError('[changelog-version] An error occurred while staging updated changelog:\n')
      logError(error)
      process.exit(1)
    }

    console.log(green(`\n[changelog-version] succesfully updated changelog to the v${version}`))
  })
})

if (isFromCLI) updateChangelogVersion()

module.exports = {
  updateVersion,
  updateChangelogVersion
}
