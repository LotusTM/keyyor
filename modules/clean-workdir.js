const { exec, execSync } = require('child_process')
const { green, red } = require('chalk')

const isFromCLI = require.main === module
const logError = (message) => console.error(red(message))

const checkIsWorkDirClean = () => {
  exec('git --version', (error) => {
    if (error) {
      console.log('[clean-workdir] No Git has been detected.\n\nSkipping clean working directory test.\n')
      return console.log(error)
    }

    exec('git status --porcelain', (error, stdout) => {
      if (error) {
        logError('[clean-workdir] an error occured:\n')
        return logError(error)
      }

      if (stdout) {
        logError('[clean-workdir] Modified files have been detected.\n')
        logError('Build task should not generate or change files outside the build folder:\n')
        logError(stdout)
        console.error(execSync('git diff --color').toString())
        return process.exit(1)
      }

      console.log(green('[clean-workdir] no modified files detected outside of the build folder. Live long and prosper'))
    })
  })
}

if (isFromCLI) checkIsWorkDirClean()

module.exports = checkIsWorkDirClean
