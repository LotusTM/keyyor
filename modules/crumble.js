const _ = require('lodash')
const { extname } = require('path')

/**
 * Explodes string path into array breadcrumb
 * @param  {string} to Relative or absolute path
 * @return {array}     Array, which consists of path's fragments
 * @example crumble('/url/to/index.html') -> `['url', 'to']`
 * @example crumble('/url/to/404.html') -> `['url', 'to', '404']`
 * @example crumble('/') -> `['index']`
*/
module.exports = (to) => {
  if (to === '/') return ['index']

  const breadcrumb = _.chain(to)
    .trimStart('/')
    .trimEnd('/')
    .replace(new RegExp(`${extname(to)}$`, 'ig'), '')
    .split('/')
    .value()

  if ((breadcrumb.length >= 2) && (_.last(breadcrumb) === 'index')) {
    breadcrumb.pop()
  }

  return breadcrumb
}
