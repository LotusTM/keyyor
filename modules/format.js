const { sprintf } = require('sprintf-js')
const { vsprintf } = require('sprintf-js')
const traverse = require('./traverse')

/**
 * Replace placeholders with provided values via `sprintf` or `vsprintf`.
 * `printf` type will be chosen depending on provided placeholders
 * @param {*}                   input           Input in which should be made replacement
 * @param {string|object|array} placeholders... List of placeholders, object with named
 *                                              placeholders or arrays of placeholders
 * @return {*} Transformed input
*/
module.exports = (input, ...placeholders) => {
  const [placeholder] = placeholders
  const useVsprintf = placeholders.length === 1 && Array.isArray(placeholder)

  return traverse(
    input,
    (tmpl) => useVsprintf ? vsprintf(tmpl, placeholder) : sprintf(tmpl, ...placeholders),
    (string) => /%([\w+-]|\.\d|[\w_]'|\([\w])/.test(string)
  )
}
