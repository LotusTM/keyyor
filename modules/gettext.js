const Gettext = require('node-gettext')
const { po } = require('gettext-parser')
const { join, extname } = require('path')
const { readFileSync } = require('fs')
const { file: { expand } } = require('grunt')

const DEFAULT_DOMAIN = 'messages'

const getLocaleDirs = (cwd) => expand({ cwd, filter: 'isDirectory' }, '*', '!*templates')

const readPoMessages = (path) => {
  const file = readFileSync(path)
  return po.parse(file)
}

const autobindTextdomains = (gt, cwd, locale) => {
  cwd = join(cwd, locale)

  expand({ cwd, filter: 'isFile' }, '{,**/}*.{mo,po}')
    .forEach(domainpath => {
      const ignored = new RegExp(`${extname(domainpath)}$|LC_MESSAGES\\/`)
      const domain = domainpath.replace(ignored, '')
      const domainMessages = readPoMessages(join(cwd, domainpath))

      gt.addTranslations(locale, domain, domainMessages)
    })
}

module.exports = (cwd) => {
  if (!cwd) {
    throw new Error('[gettext] argument `options.cwd` should be privided')
  }

  const gt = new Gettext()

  getLocaleDirs(cwd).forEach((localeDir) => {
    autobindTextdomains(gt, cwd, localeDir)
  })

  gt.on('error', (error) => console.error(
    `[gettext] ${error}`
  ))

  return {
    setLocale: gt.setLocale,
    setTextdomain: gt.setTextDomain,
    on: gt.on,
    off: gt.off,
    gettext: gt.gettext,
    dgettext: gt.dgettext,
    ngettext: gt.ngettext,
    dngettext: gt.dngettext,
    pgettext: gt.pgettext,
    dpgettext: gt.dpgettext,
    npgettext: gt.npgettext,
    dnpgettext: gt.dnpgettext,

    nunjucksExtensions (env, currentLocale) {
      gt.setLocale(currentLocale)
      gt.setTextDomain(DEFAULT_DOMAIN)

      env.addGlobal('setLocale', (locale = currentLocale) => gt.setLocale.bind(gt)(locale))
      env.addGlobal('setTextdomain', (domain = DEFAULT_DOMAIN) => gt.setTextDomain.bind(gt)(domain))
      env.addGlobal('gettext', gt.gettext.bind(gt))
      env.addGlobal('dgettext', gt.dgettext.bind(gt))
      env.addGlobal('ngettext', gt.ngettext.bind(gt))
      env.addGlobal('dngettext', gt.dngettext.bind(gt))
      env.addGlobal('pgettext', gt.pgettext.bind(gt))
      env.addGlobal('dpgettext', gt.dpgettext.bind(gt))
      env.addGlobal('npgettext', gt.npgettext.bind(gt))
      env.addGlobal('dnpgettext', gt.dnpgettext.bind(gt))
    }
  }
}
