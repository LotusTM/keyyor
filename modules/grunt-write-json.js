const { cyan } = require('chalk')

module.exports = ({ registerMultiTask, log, verbose, file: { write }, util: { pluralize } }) =>
  registerMultiTask('writeJSON', 'Write a JSON-file', function () {
    const { dataFn } = this.options()

    if (typeof dataFn !== 'function') throw new Error(`\`dataFn\` should be a function which returns data`)
    if (!this.files.length) return log.error('No files specified.')

    const data = dataFn()

    this.files.forEach(({ dest }) => {
      if (!dest) return log.error('No dest file specified')

      verbose.ok(`\nProcessed: ${cyan(dest)}`)

      const exps = Object.keys(data).reduce((exps, key) => {
        return `${exps}\nexport const ${key} = ${JSON.stringify(data[key], null, 2)}`
      }, '')

      const expDefaults = Object.keys(data).reduce((exps, key) => `${exps}${key},\n`, '')

      write(dest, `${exps}\n\nexport default {\n${expDefaults}}`)
      verbose.ok(`File ${cyan(dest)} created`)
    })

    log.ok(`${cyan(this.files.length)} ${pluralize(this.files.length, 'file/files')} processed`)
  })
