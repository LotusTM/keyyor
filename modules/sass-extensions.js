const sass = require('node-sass')
const { castToSass } = require('node-sass-utils')(sass)
const { get } = require('lodash')
const onecolor = require('onecolor')

/**
 * Get property from data. If returned property turns out
 * to be a color represented as a string, it will be cast to a
 * Sass color
 * @param {object}       data  Data to work with
 * @param {array|string} query Query to property in data, according to https://lodash.com/docs#get
 * @return {*} Requested property
 * @example
 *   $images-path: '/' + kotsu-data('PATH.images')
 *   $primary-color: kotsu-data('SITE.themeColor')
*/
const kotsuData = (data) => (query) => {
  const value = get(data, query.getValue())
  const color = onecolor(value)

  if (color) {
    return sass.types.Color(
      color.red() * 255,
      color.green() * 255,
      color.blue() * 255,
      color.alpha()
    )
  }

  return castToSass(value)
}

module.exports = {
  kotsuData,

  /**
   * Pass Sass functions to `node-sass`
   * @param {object} data Data to be accessible in Sass
   * @return {object} Sass functions
   */
  functions: (data) => ({
    'kotsu-data($query)': kotsuData(data)
  })
}
