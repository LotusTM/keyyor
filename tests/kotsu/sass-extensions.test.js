/* eslint-env jest */

import sass from 'node-sass'
import utils from 'node-sass-utils'
import { kotsuData } from '../../modules/sass-extensions'

const { castToSass } = utils(sass)

const data = {
  one: {
    oneOne: 'level2 value'
  },
  two: {
    twoOneColor: '#fff'
  }
}

describe('Sass extension', () => {
  describe('kotsuData (kotsu-data)', () => {
    it('should get from data', () => {
      expect(kotsuData(data)(castToSass('one.oneOne')).getValue()).toMatchSnapshot()
    })

    it('should cast colors to Sass color', () => {
      const c = kotsuData(data)(castToSass('two.twoOneColor'))

      expect([
        c.getR(),
        c.getB(),
        c.getG(),
        c.getA()
      ]).toMatchSnapshot()
    })
  })
})
