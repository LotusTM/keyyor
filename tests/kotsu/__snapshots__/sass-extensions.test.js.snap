// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sass extension kotsuData (kotsu-data) should cast colors to Sass color 1`] = `
Array [
  255,
  255,
  255,
  1,
]
`;

exports[`Sass extension kotsuData (kotsu-data) should get from data 1`] = `"level2 value"`;
